plugins/timertop/timermodel.cpp

```diff
@@ -27,190 +27,387 @@
+void TimerModel::pushChanges()
 {
-    // First, return the timer info if it already exists
-    foreach (const TimerInfoPtr &freeTimer, m_freeTimers) {
-        if (freeTimer->timerId() == timerId)
-            return freeTimer;
+    if (!TimerModel::isInitialized())
+        return;
+
+    QMutexLocker locker(Probe::objectLock());
+    TimerIdInfoHash infoHash;
+
+    infoHash.reserve(s_gatheredTimersData.count());
+    for (auto it = s_gatheredTimersData.begin(), end = s_gatheredTimersData.end(); it != end; ++it) {
+        TimerIdData &itInfo = it.value();
+
+        if (itInfo.changed) {
+            infoHash.insert(it.key(), itInfo.toInfo(it.key().type()));
+            itInfo.changed = false;
+        }
     }
+    infoHash.squeeze();
 
-    // Create a new free timer, and emit the correct update signals
-    TimerInfoPtr timerInfo(new TimerInfo(timerId));
-    beginInsertRows(QModelIndex(), rowCount(), rowCount());
-    m_freeTimers.append(timerInfo);
-    endInsertRows();
-    return timerInfo;
+    s_timerModel->m_applyChangesMethod.invoke(s_timerModel, Qt::QueuedConnection,
+                                              Q_ARG(GammaRay::TimerIdInfoHash, infoHash));
 }
 
-TimerInfoPtr TimerModel::findOrCreateQTimerTimerInfo(QObject *timer)
+bool TimerModel::eventNotifyCallback(void *data[])
 {
-    if (!timer)
-        return TimerInfoPtr();
-
-    QVariant timerInfoVariant = timer->property(timerInfoPropertyName);
-    if (!timerInfoVariant.isValid()) {
-        const TimerInfoPtr info = TimerInfoPtr(new TimerInfo(timer));
-        if (m_qmlTimerTriggeredIndex < 0 && info->type() == TimerInfo::QQmlTimerType)
-            m_qmlTimerTriggeredIndex = timer->metaObject()->indexOfMethod("triggered()");
-        timerInfoVariant.setValue(info);
-        if (timer->thread() == QThread::currentThread()) // ### FIXME: we shouldn't use setProperty() in the first place...
-            timer->setProperty(timerInfoPropertyName, timerInfoVariant);
+    /*
+     * data[0] == receiver
+     * data[1] == event
+     * data[2] == bool result ref, what is returned by caller if this function return true
+     * (QCoreApplication::notifyInternal / QCoreApplication::notifyInternal2)
+     */
+    QObject *receiver = static_cast<QObject *>(data[0]);
+    QEvent *event = static_cast<QEvent *>(data[1]);
+//    bool *result = static_cast<bool *>(data[2]);
+
+    if (event->type() == QEvent::Timer) {
+        const QTimerEvent *const timerEvent = static_cast<QTimerEvent *>(event);
+        Q_ASSERT(timerEvent->timerId() != -1);
+        const QTimer *const timer = qobject_cast<QTimer *>(receiver);
+
+        // If there is a QTimer associated with this timer ID, don't handle it here, it will be handled
+        // by the signal hooks preSignalActivate/postSignalActivate.
+        if (timer && timer->timerId() == timerEvent->timerId()) {
+            return false;
+        }
+
+        {
+            QMutexLocker locker(Probe::objectLock());
```

Comment:
I guess the lock is needed to protect s_gatheredTimersData? In that case we probably should use our own lock here rather than abuse the probe's object lock?